File: compile.py
from json import JSONEncoder, dumps
from dataclasses import dataclass, is_dataclass, asdict
from typing import Optional, Literal

from locations import items, locations


class DataClassJSONEncoder(JSONEncoder):
    def default(self, o):
        if is_dataclass(o):
            return self.clean_nones(asdict(o))
        return super().default(o)

    @classmethod
    def clean_nones(cls, value):
        """
        Recursively remove all None values from dictionaries and lists, and returns
        the result as a new dictionary or list.
        """
        if isinstance(value, list):
            return [cls.clean_nones(x) for x in value if x is not None]
        elif isinstance(value, dict):
            return {key: cls.clean_nones(val) for key, val in value.items() if val is not None}
        else:
            return value


@dataclass
class StageDefinition:
    name: str
    img: Optional[str] = None
    img_mods: Optional[str] = None
    disabled_img_mods: Optional[str] = None
    disabled_img: Optional[str] = None
    codes: Optional[str] = None
    inherit_codes: Optional[bool] = None


@dataclass
class ItemDefinition:
    name: str
    type: Literal[
        "static",
        "progressive",
        "toggle",
        "consumable",
        "progressive_toggle",
        "composite_toggle",
        "toggle_badged",
    ]
    img: Optional[str] = None
    img_mods: Optional[str] = None
    disabled_img: Optional[str] = None
    disabled_img_mods: Optional[str] = None
    codes: Optional[str] = None
    stages: Optional[list[StageDefinition]] = None
    initial_stage_idx: Optional[int] = None
    loop: Optional[bool] = None
    allow_disabled: Optional[bool] = None


@dataclass
class MapLocationDefinition:
    map: str
    x: int
    y: int


@dataclass
class LocationDefinition:
    name: str
    chest_opened_img: str
    chest_unopened_img: str
    access_rules: str | list[str] = ""
    visibility_rules: str | list[str] = ""
    map_locations: Optional[list[MapLocationDefinition]] = None


from entrances import entrances


er_items: list[ItemDefinition] = [
    *[
        ItemDefinition(
            name=f"{entrance.name} - Entrance",
            type="toggle",
            img=f"images/er_legend/er_{entrance.code}_ent.png",
            disabled_img_mods="none",
            codes=f"__er_{entrance.code}_ent",
        )
        for entrance in entrances
    ],
    *[
        ItemDefinition(
            name=f"{entrance.name} - Destination Node",
            type="progressive",
            loop=True,
            codes=f"__er_{entrance.code}_dst",
            initial_stage_idx=0,
            allow_disabled=False,
            stages=[
                StageDefinition(
                    name="Unknown Stage Destination",
                    inherit_codes=False,
                    img="images/er_legend/er_unknown_dst.png",
                ),
                *[
                    StageDefinition(
                        name=f"{inner_entrance.name} - Destination",
                        inherit_codes=False,
                        img=f"images/er_legend/er_{inner_entrance.code}_dst.png",
                    )
                    for inner_entrance in entrances
                ],
            ],
        )
        for entrance in entrances
    ],
]


def compile_all():
    with open("items/er_items.json", "w") as file:
        file.write("// This file is auto-generated. Do not make modifications to this file directly.\n")
        file.write(dumps(er_items, cls=DataClassJSONEncoder))

    with open("items/location_items.json", "w") as file:
        file.write("// This file is auto-generated. Do not make modifications to this file directly.\n")
        file.write(dumps(items))

    with open("scripts/autotracking/location_mapping.lua", "w") as file:
        file.write("-- This file is auto-generated. Do not make modifications to this file directly.\n")
        file.write("LOCATION_MAPPING = {\n")
        for location in locations:
            if location.code == 0:
                continue

            file.write(f'    [{location.code}] = {{"{location.get_code_name()}"}},\n')
        file.write("}")

    with open(f"locations/entrances.json", "w") as file:
        file.write("// This file is auto-generated. Do not make modifications to this file directly.\n")
        region_entrances = [
            {
                "name": f"{entrance.name} Entrance",
                "access_rules": entrance.access_rules,
                "children": [
                    {
                        "name": "Unknown Destination",
                        "visibility_rules": f"$CanNotSee|{entrance.code}",
                        "access_rules": [rule for rule in entrance.access_rules],
                        "map_locations": [
                            {
                                "map": "map_castle",
                                "x": entrance.coords[0],
                                "y": entrance.coords[1],
                            }
                        ],
                        "sections": [
                            {
                                "name": "Enter Stage and Set Entrance",
                                "hosted_item": "__unknown_er",
                            }
                        ],
                    },
                    {
                        "name": "Bowser in the Sky",
                        "access_rules": ["$CanAccessUpstairs,$HasStars|F2Door,$HasStars|F3Door"],
                        "map_locations": [{"map": "map_castle", "x": 1366, "y": 520}],
                        "sections": [
                            {
                                "name": location.name,
                                "hosted_item": location.get_code_name(),
                                "access_rules": location.access_rules,
                                "visibility_rules": location.visibility_rules(),
                            }
                            for location in locations
                            if location.region == "Bowser in the Sky"
                        ],
                    },
                    *[
                        {
                            "name": region.stageify(),
                            "map_locations": [
                                {
                                    "map": "map_castle",
                                    "x": entrance.coords[0],
                                    "y": entrance.coords[1],
                                }
                            ],
                            "visibility_rules": f"$CanSee|{region.code}|{entrance.code}",
                            "sections": [
                                {
                                    "name": location.name,
                                    "hosted_item": location.get_code_name(),
                                    "access_rules": location.access_rules,
                                    "visibility_rules": location.visibility_rules(),
                                }
                                for location in locations
                                if location.in_region(region.name)
                            ],
                        }
                        for region in entrances
                    ],
                ],
            }
            for entrance in entrances
        ]

        file.write(dumps(region_entrances))

File: packager/compile.ts
import { mkdir, readdir } from "node:fs/promises";

import { parse as parseYAML } from "yaml";

import { Entrance, EntranceData } from "./entrance.ts";
import { Location, LocationData } from "./location.ts";
import { Region, RegionData } from "./region.ts";

const areas = {
    BoB: "Bob-omb Battlefield",
    WF: "Whomp's Fortress",
    JRB: "Jolly Roger Bay",
    CCM: "Cool, Cool Mountain",
    BBH: "Big Boo's Haunt",
    HMC: "Hazy Maze Cave",
    LLL: "Lethal Lava Land",
    SSL: "Shifting Sand Land",
    DDD: "Dire, Dire Docks",
    SL: "Snowman's Land",
    WDW: "Wet-Dry World",
    TTM: "Tall, Tall Mountain",
    THI: "Tiny-Huge Island",
    TTC: "Tick Tock Clock",
    RR: "Rainbow Ride",
    BitDW: "Bowser in the Dark World",
    BitFS: "Bowser in the Fire Sea",
    BitS: "Bowser in the Sky",
    TotWC: "Tower of the Wing Cap",
    CotMC: "Cavern of the Metal Cap",
    VCutM: "Vanish Cap under the Moat",
    PSS: "Princess's Secret Slide",
    SA: "Secret Aquarium",
    WMotR: "Wing Mario over the Rainbow",
    MKC: "Mushroom Kingdom Castle",
} as const;

async function readCSV<T>(cwd: string, path: string): Promise<T[]> {
    const buffer = await Bun.file(`${cwd}/${path}`).text();
    const rows: T[] = [];
    for (const row of buffer.trim().split("\n").slice(1)) {
        rows.push(row
            .split(/(?<!\\),/)
            .map((d) => d.replaceAll("\\,", ","))
            .map((d) => d.replaceAll("\"", "")) as T,
        );
    }

    return rows;
}

async function compileStaticData(cwd: string): Promise<void> {
    const dirs = ["items", "layouts", "locations", "maps"];
    for (const dir of dirs) {
        const inPath = `${cwd}/data/${dir}`;
        const outPath = `${cwd}/build/${dir}`;
        try {
            await mkdir(outPath);
        } catch { /* If dir exists, do nothing. */ }

        for (const file of await readdir(inPath)) {
            console.log(`\tCompiling 'data/${dir}/${file}'`);
            const buffer = await Bun.file(`${inPath}/${file}`).text();
            const obj = parseYAML(buffer) as { data: unknown };
            await Bun.write(`${outPath}/${file.split(".")[0]}.json`,
                "/* This file was automatically generated via Phakager. Do not make manual edits to this file. */\n"
                + JSON.stringify(obj.data, null, 4),
            );
        }
    }
}

async function compileAreaRandomizerItems(cwd: string): Promise<void> {
    const entranceItems = [];
    const destinationItems = [];
    for (const entrance of Entrance.entrances) {
        entranceItems.push({
            name: `${entrance.name} - Entrance`,
            type: "toggle",
            img: `images/er_legend/er_${entrance.acronym.toLowerCase()}_ent.png`,
            disabled_img_mods: "none",
            codes: `__er_${entrance.acronym}_ent`,
        });

        const destinationItem = {
            name: `${entrance.name} - Destination Symbol`,
            type: "progressive",
            loop: true,
            initial_stage_idx: entrance.acronym === "BitS" ? entrance.order : 0,
            allow_disabled: false,
            codes: `__er_${entrance.acronym}_dst`,
            stages: [{
                name: "??? - Destination",
                img: "images/er_legend/er_unknown_dst.png",
                inherit_codes: false,
            }],
        };

        for (const destination of Entrance.entrances) {
            destinationItem.stages.push({
                name: `${destination.name} - Destination`,
                img: `images/er_legend/er_${destination.acronym.toLowerCase()}_dst.png`,
                inherit_codes: false,
            });
        }

        destinationItems.push(destinationItem);
    }

    console.log("\tBuilding 'items/area_rando.json'");
    const output = JSON.stringify([...entranceItems, ...destinationItems], null, 4);
    await Bun.write(`${cwd}/build/items/area_rando.json`,
        "/* This file was automatically generated via Phakager. Do not make manual edits to this file. */\n" + output,
    );
}

async function compileLocationItems(cwd: string): Promise<void> {
    const locationItems = [];
    locationItems.push({
        name: "Enter Stage and Set Entrance",
        type: "toggle",
        codes: "__location_item_null",
    });

    for (const location of Location.locations) {
        locationItems.push({
            name: location.name,
            type: "toggle",
            codes: location.itemCode,
            img: location.icons[1],
            disabled_img: location.icons[0],
            disabled_img_mods: "none",
        });
    }

    console.log("\tBuilding 'items/locations.json'");
    const output = JSON.stringify(locationItems, null, 4);
    await Bun.write(`${cwd}/build/items/locations.json`,
        "/* This file was automatically generated via Phakager. Do not make manual edits to this file. */\n" + output,
    );
}

async function compileCastleMapEntranceLocations(cwd: string): Promise<void> {
    const overworldEntrances: unknown[] = [];
    for (const entrance of Entrance.entrances) {
        const entranceNode: { [p: string]: unknown, children: unknown[] } = {
            name: `${entrance.name} Entrance`,
            access_rules: entrance.accessRules,
            children: [
                {
                    name: "Unknown Stage",
                    visibility_rules: `$IsUnknownDestination|${entrance.acronym}`,
                    access_rules: entrance.accessRules,
                    map_locations: [{
                        map: "map_castle",
                        x: entrance.coords[0],
                        y: entrance.coords[1],
                    }],
                    sections: [{
                        name: "Enter Stage and Set Entrance",
                        hosted_item: "__location_item_null",
                    }],
                },
                {
                    name: "Unknown Stage [Scout]",
                    visibility_rules: `$IsUnknownDestination|${entrance.acronym}`,
                    access_rules: `{$IsUnknownDestination|${entrance.acronym}}`,
                    map_locations: [{
                        map: "map_castle",
                        x: entrance.coords[0],
                        y: entrance.coords[1],
                    }],
                    sections: [{
                        name: "Enter Stage and Set Entrance",
                        hosted_item: "__location_item_null",
                    }],
                },
            ],
        };

        const scoutNode = {
            name: "[Z] Entrance Accessibility",
            access_rules: [...entrance.accessRules, ...entrance.scoutRules.map((i) => `{${i}}`)],
            visibility_rules: ["$AreaRando"],
            children: [{
                name: `Entrance Accessibility for ${entrance.name}`,
                map_locations: [{
                    map: "map_castle",
                    x: entrance.coords[0] + 16,
                    y: entrance.coords[1] + 16,
                    size: 16,
                    shape: "trapezoid",
                }],
                sections: [{
                    name: "Green: Accessible\n"
                        + "Yellow: Sequence Break Required\n"
                        + "Blue: Potentially Scoutable (i.e., TTC hands)\n"
                        + "Red: Logically Inaccessible",
                    hosted_item: `__er_${entrance.acronym}_dst`,
                }],
            }],
        };

        for (const [acronym, stage] of Object.entries(areas)) {
            const locationNode: { [p: string]: unknown, sections: unknown[] } = {
                name: stage,
                visibility_rules: `$IsSelectedDestination|${acronym}|${entrance.acronym}`,
                map_locations: [{
                    map: "map_castle",
                    x: entrance.coords[0],
                    y: entrance.coords[1],
                }],
                sections: [],
            };

            for (const location of Location.locations.filter((l) => l.region.stage === acronym)) {
                locationNode.sections.push({
                    name: location.name,
                    hosted_item: location.itemCode,
                    access_rules: location.accessRules,
                    visibility_rules: location.visibilityRules,
                });
            }

            entranceNode.children.push(locationNode);
        }

        overworldEntrances.push(entranceNode);
        overworldEntrances.push(scoutNode);
    }

    console.log("\tBuilding 'locations/castle_entrances.json'");
    const output = JSON.stringify(overworldEntrances, null, 4);
    await Bun.write(`${cwd}/build/locations/castle_entrances.json`,
        "/* This file was automatically generated via Phakager. Do not make manual edits to this file. */\n" + output,
    );
}

async function compileLocationMappingLua(cwd: string): Promise<void> {
    let output = "LOCATION_MAPPING = {\n";
    for (const location of Location.locations) {
        output += `    [${location.code}] = {"${location.itemCode}"}, -- ${location.region.stage.padEnd(5)} - ${location.name}\n`;
    }

    console.log("\tBuilding 'scripts/autotracking/location_mapping.lua'");
    await Bun.write(`${cwd}/build/scripts/autotracking/location_mapping.lua`,
        "-- This file was automatically generated via Phakager. Do not make manual edits to this file.\n" + output + "}",
    );
}

export async function compileAll(cwd: string): Promise<void> {
    const entrances = await readCSV<EntranceData>(cwd, "data/entrances.csv");
    const regions = await readCSV<RegionData>(cwd, "data/regions.csv");
    const locations = await readCSV<LocationData>(cwd, "data/locations.csv");

    // Pre-build entrances, regions, and locations.
    entrances.forEach((entrance) => Entrance.create(entrance));
    for (const stage of Object.keys(areas)) {
        Region.create([stage, null, ""]);
    }

    regions.forEach((region) => Region.create(region));
    locations.forEach((location) => Location.create(location));

    // Compile everything.
    await compileStaticData(cwd);
    await compileAreaRandomizerItems(cwd);
    await compileLocationItems(cwd);
    await compileCastleMapEntranceLocations(cwd);
    await compileLocationMappingLua(cwd);
}
